File: src/main/kotlin/org/alexn/hook/CommandTrigger.kt
package org.alexn.hook

import arrow.core.Either
import arrow.core.left
import arrow.core.right
import kotlinx.coroutines.TimeoutCancellationException
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.withTimeout
import java.io.File
import java.util.concurrent.atomic.AtomicReference
import kotlin.time.Duration.Companion.seconds

/**
 * Handles the actual shell command execution, per project.
 */
class CommandTrigger private constructor(
    private val projects: Map<String, AppConfig.Project>,
    private val locks: AtomicReference<Map<String, Mutex>>,
) {
    private fun lockFor(key: String): Mutex {
        val lockRef = locks.get()[key]
        if (lockRef != null) return lockRef

        val updatedMap =
            locks.updateAndGet { currentMap ->
                val r = currentMap[key]
                if (r != null) return@updateAndGet currentMap

                val newRef = Mutex()
                val newMap = currentMap + mapOf(key to newRef)
                newMap
            }
        return updatedMap[key]!!
    }

    suspend fun triggerCommand(key: String): Either<RequestError, Unit> {
        val project =
            projects[key]
                ?: return RequestError.NotFound("Project `$key` does not exist").left()

        val timeoutDuration = project.timeout ?: 30.seconds
        val mutex = lockFor(key)
        mutex.lock()
        return try {
            val result =
                withTimeout(timeoutDuration) {
                    executeRawShellCommand(
                        command = project.command,
                        dir = File(project.directory),
                    )
                }
            if (result.isSuccessful) {
                Unit.right()
            } else {
                RequestError.Internal(
                    "Command execution failed",
                    null,
                    meta =
                        mapOf(
                            "exit-code" to result.exitCode.toString(),
                            "stdout" to result.stdout,
                            "stderr" to result.stderr,
                        ),
                ).left()
            }
        } catch (e: TimeoutCancellationException) {
            RequestError.TimedOut(
                "Command execution timed-out after $timeoutDuration",
            ).left()
        } finally {
            mutex.unlock()
        }
    }

    companion object {
        /**
         * Builder with side effects.
         */
        operator fun invoke(projects: Map<String, AppConfig.Project>): CommandTrigger =
            CommandTrigger(
                projects,
                AtomicReference(mapOf()),
            )
    }
}

File: src/main/scala/org/alexn/hook/CommandTrigger.scala
package org.alexn.hook

import cats.data.EitherT
import cats.effect.kernel.{Async, Sync}
import cats.effect.std.Semaphore
import cats.effect.syntax.all.*
import cats.syntax.all.*

import java.util.concurrent.atomic.AtomicReference
import scala.concurrent.TimeoutException
import scala.concurrent.duration.*

final class CommandTrigger[F[_]](
    projects: Map[String, ProjectConfig],
    locks: AtomicReference[Map[String, Semaphore[F]]]
)(using
    Async[F],
    OperatingSystem[F]
):
    def lockFor(key: String): F[Semaphore[F]] =
        locks.modifyDelayed: locks =>
            locks.get(key) match
            case Some(mutex) => Right(mutex)
            case None =>
                Left(Semaphore[F](1).map: mutex =>
                    (locks.updated(key, mutex), mutex))

    def triggerCommand(key: String): EitherT[F, RequestError, Unit] =
        for
            project <- EitherT.fromOption(
                projects.get(key),
                RequestError.NotFound(s"Project not found: $key")
            )
            result <- EitherT:
                val timeoutDuration = project.timeout.getOrElse(30.seconds)
                val task =
                    for
                        mutex <- lockFor(key)
                        result <-
                            mutex.permit.use: _ =>
                                OperatingSystem[F].executeRawShellCommand(
                                    project.command,
                                    dir = Some(project.directory)
                                )
                            .timeout(timeoutDuration)
                    yield
                        if result.exitCode == 0 then
                            Right(())
                        else
                            Left(RequestError.InternalServerError(
                                s"Command execution failed for `$key``",
                                cause = null,
                                meta = Map(
                                    "exit-code" -> result.exitCode.toString,
                                    "stdout" -> result.stdout,
                                    "stderr" -> result.stderr
                                )
                            ))
                task.recover:
                    case _: TimeoutException =>
                        Left(RequestError.TimedOut(
                            s"Command execution timed-out after $timeoutDuration"
                        ))
        yield result
end CommandTrigger

object CommandTrigger:
    def apply[F[_]](projects: Map[String, ProjectConfig])(using
        Async[F],
        OperatingSystem[F]
    ): F[CommandTrigger[F]] =
        Sync[F].delay:
            val ref = new AtomicReference[Map[String, Semaphore[F]]](Map.empty)
            new CommandTrigger[F](projects, ref)
